pkg/encryption/encryption.go

```diff
@@ -0,0 +1,227 @@
+package encryption
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"text/template"
+	"time"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	capi "sigs.k8s.io/cluster-api/api/v1alpha3"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/encryption-provider-operator/pkg/key"
+	"github.com/giantswarm/encryption-provider-operator/pkg/label"
+	"github.com/giantswarm/encryption-provider-operator/pkg/project"
+)
+
+const (
+	EncryptionProviderConfig = "encryption"
+
+	// LegacyEncryptionProviderType is old encryption provider used  by legacy giantswarm product
+	LegacyEncryptionProviderType = "aescbc"
+
+	// DefaultEncryptionProviderType value is used for new keys
+	// https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/#providers
+	DefaultEncryptionProviderType = "secretbox"
+
+	// Poly1305KeyLength represents the 32 bytes length for Poly1305
+	// padding encryption key.
+	Poly1305KeyLength = 32
+)
+
+type Config struct {
+	Cluster                  *capi.Cluster
+	DefaultKeyRotationPeriod time.Duration
+
+	CtrlClient ctrlclient.Client
+	Logger     logr.Logger
+}
+
+type Service struct {
+	cluster *capi.Cluster
+
+	defaultKeyRotationPeriod time.Duration
+
+	ctrlClient ctrlclient.Client
+	logger     logr.Logger
+}
+
+func New(c Config) (*Service, error) {
+	if c.Cluster == nil {
+		return nil, errors.New("cluster cannot be nil")
+	}
+	if c.CtrlClient == nil {
+		return nil, errors.New("ctrlClient cannot be nil")
+	}
+	if c.Logger == nil {
+		return nil, errors.New("logger cannot be nil")
+	}
+
+	s := &Service{
+		cluster:                  c.Cluster,
+		defaultKeyRotationPeriod: c.DefaultKeyRotationPeriod,
+		ctrlClient:               c.CtrlClient,
+		logger:                   c.Logger,
+	}
+
+	return s, nil
+}
+
+func (s *Service) Reconcile() error {
+	ctx := context.TODO()
+
+	var encryptionProviderSecret v1.Secret
+
+	err := s.ctrlClient.Get(ctx, ctrlclient.ObjectKey{
+		Name:      key.SecretName(s.cluster.ClusterName),
+		Namespace: s.cluster.Namespace,
+	}, &encryptionProviderSecret)
+
+	if apierrors.IsNotFound(err) {
+		// create new encryption secret
+		err := s.createNewEncryptionProviderSecret(ctx)
+		if err != nil {
+			s.logger.Error(err, "failed to get encryption provider config secret for cluster")
+			return err
+		}
+	} else if err != nil {
+		s.logger.Error(err, "failed to get encryption provider config secret for cluster")
+		return err
+	} else {
+		// config already exists, check for key rotation
+		err = s.keyRotation(ctx)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
```

Comment:
```suggestion

	// config already exists, check for key rotation
	return s.keyRotation(ctx)
```

Reply:
Ah, didn't realise there was more to go here eventually. Worth dropping in a `// TODO:` ?